File: android/core/src/main/kotlin/com/iblocker/core/lists/BlocklistCompiler.kt
package com.iblocker.core.lists

import com.iblocker.core.rules.CompiledBlocklist
import com.iblocker.core.rules.DomainValidator
import com.iblocker.core.rules.FilterListParser
import com.iblocker.core.rules.Fnv1a
import com.iblocker.core.rules.SeedRules
import com.iblocker.core.shared.AppPaths

data class CompileStats(
    var blockedEntryCount: Int = 0,
    var listAllowEntryCount: Int = 0,
    var userAllowEntryCount: Int = 0,
    var userDenyEntryCount: Int = 0,
    var skippedLines: Int = 0,
    var perSourceCounts: MutableMap<String, Int> = LinkedHashMap(),
    var generation: UInt = 0u,
)

/**
 * Turns cached list texts + user allow/deny into the three compiled blobs the
 * service mmaps. Runs in the app process, where memory is plentiful — never
 * on the packet path.
 */
object BlocklistCompiler {

    fun compile(state: FilterListState, paths: AppPaths): CompileStats {
        val stats = CompileStats()
        val blockHashes = HashSet<ULong>()
        val listAllowHashes = HashSet<ULong>()

        paths.ensureDirectories()

        for (source in state.sources.filter { it.enabled }) {
            val text = SeedRules.bundledText(source.id) ?: run {
                val file = paths.cachedListFile(source.id)
                if (!file.isFile) return@run null
                try {
                    file.readText()
                } catch (_: Exception) {
                    null
                }
            } ?: continue

            val parsed = FilterListParser.parse(text)
            stats.skippedLines += parsed.skippedLines
            stats.perSourceCounts[source.id] = parsed.blockDomains.size
            state.metadata[source.id] = (state.metadata[source.id] ?: FilterListMetadata())
                .copy(entryCount = parsed.blockDomains.size)

            for (domain in parsed.blockDomains) blockHashes.add(Fnv1a.hash64(domain))
            for (domain in parsed.allowDomains) listAllowHashes.add(Fnv1a.hash64(domain))
        }

        // List-level @@ allows are exact-hash subtractions; user allows stay
        // separate so the service can honor suffix-level allows at runtime.
        blockHashes.removeAll(listAllowHashes)

        val userAllowHashes = normalizeToHashes(state.userAllowlist)
        val userDenyHashes = normalizeToHashes(state.userDenylist)

        state.generation += 1u
        state.compiledSeedVersion = SeedRules.VERSION
        val generation = state.generation

        CompiledBlocklist.write(blockHashes, generation, paths.blocklistFile)
        CompiledBlocklist.write(userAllowHashes, generation, paths.userAllowlistFile)
        CompiledBlocklist.write(userDenyHashes, generation, paths.userDenylistFile)

        stats.blockedEntryCount = blockHashes.size
        stats.listAllowEntryCount = listAllowHashes.size
        stats.userAllowEntryCount = userAllowHashes.size
        stats.userDenyEntryCount = userDenyHashes.size
        stats.generation = generation
        return stats
    }

    private fun normalizeToHashes(domains: List<String>): Set<ULong> {
        val hashes = HashSet<ULong>()
        for (raw in domains) {
            DomainValidator.normalize(raw)?.let { hashes.add(Fnv1a.hash64(it)) }
        }
        return hashes
    }
}

File: IBlockerKit/Sources/IBlockerKit/Lists/BlocklistCompiler.swift
import Foundation

public struct CompileStats: Sendable, Equatable {
    public var blockedEntryCount: Int = 0
    public var listAllowEntryCount: Int = 0
    public var userAllowEntryCount: Int = 0
    public var userDenyEntryCount: Int = 0
    public var skippedLines: Int = 0
    public var perSourceCounts: [String: Int] = [:]
    public var generation: UInt32 = 0

    public init() {}
}

/// Turns cached list texts + user allow/deny into the three compiled blobs
/// the tunnel mmaps. Runs in the app process, where memory is plentiful.
public enum BlocklistCompiler {

    @discardableResult
    public static func compile(state: inout FilterListState, paths: AppGroupPaths) throws -> CompileStats {
        var stats = CompileStats()
        var blockHashes = Set<UInt64>()
        var listAllowHashes = Set<UInt64>()

        // A hidden source must not keep filtering from behind the curtain: a
        // relay list enabled before the UI was hidden is skipped, not compiled.
        for source in state.sources where source.enabled && !FeatureFlags.isHidden(sourceID: source.id) {
            let text: String
            if let bundled = SeedRules.bundledText(for: source.id) {
                text = bundled
            } else {
                let url = paths.cachedListURL(sourceID: source.id)
                guard let data = try? Data(contentsOf: url),
                      let cached = String(data: data, encoding: .utf8) else { continue }
                text = cached
            }

            let parsed = FilterListParser.parse(text)
            stats.skippedLines += parsed.skippedLines
            stats.perSourceCounts[source.id] = parsed.blockDomains.count
            var metadata = state.metadata[source.id] ?? FilterListMetadata()
            metadata.entryCount = parsed.blockDomains.count
            state.metadata[source.id] = metadata

            for domain in parsed.blockDomains {
                blockHashes.insert(FNV1a.hash64(domain))
            }
            for domain in parsed.allowDomains {
                listAllowHashes.insert(FNV1a.hash64(domain))
            }
        }

        // List-level @@ allows are exact-hash subtractions; user allows stay
        // separate so the tunnel can honor suffix-level allows at runtime.
        blockHashes.subtract(listAllowHashes)

        let userAllowHashes = normalizeToHashes(state.userAllowlist)
        let userDenyHashes = normalizeToHashes(state.userDenylist)

        state.generation &+= 1
        state.compiledSeedVersion = SeedRules.version
        let generation = state.generation

        try CompiledBlocklist.write(hashes: blockHashes, generation: generation, to: paths.blocklistURL)
        try CompiledBlocklist.write(hashes: userAllowHashes, generation: generation, to: paths.userAllowlistURL)
        try CompiledBlocklist.write(hashes: userDenyHashes, generation: generation, to: paths.userDenylistURL)

        stats.blockedEntryCount = blockHashes.count
        stats.listAllowEntryCount = listAllowHashes.count
        stats.userAllowEntryCount = userAllowHashes.count
        stats.userDenyEntryCount = userDenyHashes.count
        stats.generation = generation
        return stats
    }

    private static func normalizeToHashes(_ domains: [String]) -> Set<UInt64> {
        var hashes = Set<UInt64>()
        for raw in domains {
            if let domain = DomainValidator.normalize(raw) {
                hashes.insert(FNV1a.hash64(domain))
            }
        }
        return hashes
    }
}
